### ISSUE TITLE
Chromedriver downloads are targeting unavailable version

### ISSUE THREAD

**Issue by hoylemd** [CircleCI-Public/browser-tools-orb#90]
### Orb version 

1.4.2

### What happened

When trying to run the `browser-tools/install-chromedriver` step, the download 404's out.

```
Installed version of Google Chrome is 116.0.5845.140 
116.0.5845.140 will be installed
curl: (22) The requested URL returned error: 404 

Exited with code exit status 22
```

That particular version is marked as the "Stable(upcoming)" version on the [availability dashboard](https://googlechromelabs.github.io/chrome-for-testing/), and as of this writing it's unavailable

![Screen Shot 2023-08-29 at 15 38 01](https://github.com/CircleCI-Public/browser-tools-orb/assets/569403/988b3748-3cc9-45c9-912c-ac40ff422c8c)

### Expected behavior

Not quite sure the idea solution, but maybe somehow checking for working versions of both? before committing to a version?

#75 seems kind of related

Edits: corrected orb version # & replaced error example screenshot with code fence for better discoverabity

**Comment by chefvivica** [comment]
> I was also able to workaround it with
> 
> ```
>     steps:
>     - browser-tools/install-browser-tools:
>         chrome-version: 116.0.5845.96 # TODO: remove when chromedriver downloads are fixed
> ```
> 
> that also installs firefox and geckodriver, but doesn't need the apt get update at least

@hoylemd does this work for 1.4.4?

**Comment by chefvivica** [comment]
> > > I was also able to workaround it with
> > > ```
> > >     steps:
> > >     - browser-tools/install-browser-tools:
> > >         chrome-version: 116.0.5845.96 # TODO: remove when chromedriver downloads are fixed
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > that also installs firefox and geckodriver, but doesn't need the apt get update at least
> > 
> > 
> > @hoylemd does this work for 1.4.4?
> 
> It does in my project

Great. It also works for me. Thanks @hoylemd

**Comment by 0xadada** [comment]
> @0xadada What version of browser-tools is that working with? I'm getting an error on 1.4.4 when I try that:

@hoylemd version `1.4.4`, i needed to add the [apt-get step](https://github.com/CircleCI-Public/browser-tools-orb/issues/90#issuecomment-1698052540) for it to work. @chefvivica try it.

**Comment by chefvivica** [comment]
> - browser-tools/install-browser-tools:
>         chrome-version: 116.0.5845.96 # TODO: remove when chromedriver downloads are fixed

@0xadada  it worked without adding it from my end.

**Comment by ryanbourdais** [comment]
Hi all, it looks like the orb is functioning as intended, but google released a new version of google chrome stable without a coinciding chromedriver, the only real fix at the moment is to do as @hoylemd suggested and specify your chrome version.

**Comment by ryanbourdais** [comment]
Hi @hoylemd the orb actually targets the coinciding chromedriver version for the downloaded chrome version, the issue arises from google changing what version is considered latest stable by google or if a version of chrome is preinstalled on the image.

**Comment by snyaggarwal** [comment]
> > > > I was also able to workaround it with
> > > > ```
> > > >     steps:
> > > >     - browser-tools/install-browser-tools:
> > > >         chrome-version: 116.0.5845.96 # TODO: remove when chromedriver downloads are fixed
> > > > ```
> > > > 
> > > > 
> > > >     
> > > >       
> > > >     
> > > > 
> > > >       
> > > >     
> > > > 
> > > >     
> > > >   
> > > > that also installs firefox and geckodriver, but doesn't need the apt get update at least
> > > 
> > > 
> > > @hoylemd does this work for 1.4.4?
> > 
> > 
> > It does in my project
> 
> Great. It also works for me. Thanks @hoylemd

This works for everything except MacOs. Looks like it ignores `chrome-version`

**Comment by timdiggins** [comment]
@ryanbourdais 

> ... the orb actually targets the coinciding chromedriver version for the downloaded chrome version, the issue arises from google changing what version is considered latest stable by google or if a version of chrome is preinstalled on the image.

Yes, agreed, that's what it does. After reading https://github.com/GoogleChromeLabs/chrome-for-testing, I don't think that the Google Chrome team aren't guaranteeing that the release version (build number) of the current Chrome stable will match the current Chromedriver stable release version (build number) - instead they seem to be offering a current Chromedriver stable that will work against the current Chrome stable. (see https://googlechromelabs.github.io/chrome-for-testing/last-known-good-versions-with-downloads.json)

And it's possible this situation will repeat from time to time after each Google chrome stable patch release (the previous stable chromedriver will work with the new stable chrome, but the new upcoming chromedriver won't yet work -- because the release cycles of chrome and chromedriver are not fully in sync).

From my intention (and I suspect others):

As a orb-user I may want to install the stable channel for Chrome and chromedriver, but don't care about version numbers. In which case, it would be idea if the orb would take an optional channel parameter ("Stable", "Beta", "Dev" or "Canary") and thenfor its installs use https://googlechromelabs.github.io/chrome-for-testing/last-known-good-versions-with-downloads.json to look up the relevant chromedriver url, rather than constructing it based on the currently installed chrome version.

I recognize this is a major addition but it would then massively improve developer OOB experience (especially if the default for both install-chrome and install-chromedriver was channel: "Stable")

**Comment by cusher** [comment]
> After reading https://github.com/GoogleChromeLabs/chrome-for-testing, I don't think that the Google Chrome team ~aren't~ are guaranteeing that the release version (build number) of the current Chrome stable will match the current Chromedriver stable release version (build number) - instead they seem to be offering a current Chromedriver stable that will work against the current Chrome stable. (see https://googlechromelabs.github.io/chrome-for-testing/last-known-good-versions-with-downloads.json)

Unless I'm misreading it, [this page](https://chromedriver.chromium.org/downloads/version-selection) seems to indicate that the CfT listed versions should be used to drive the version of both Chrome and ChromeDriver, with a matching version:

> Starting with M115 the ChromeDriver release process is integrated with that of Chrome. The latest Chrome + ChromeDriver releases per release channel (Stable, Beta, Dev, Canary) are available at[ the Chrome for Testing (CfT) availability dashboard](https://googlechromelabs.github.io/chrome-for-testing/). As a result, you might no longer have a need for version selection — you could choose any available CfT version and simply download the correspondingly-versioned ChromeDriver binary.

**Comment by vietqhoang** [comment]
> I was also able to workaround it with
> 
> ```
>     steps:
>     - browser-tools/install-browser-tools:
>         chrome-version: 116.0.5845.96 # TODO: remove when chromedriver downloads are fixed
> ```
> 
> that also installs firefox and geckodriver, but doesn't need the apt get update at least

I was still having issues with ChromeDriver still trying to install version 116.0.5845.140, even though 116.0.5845.96 was explicitly defined.

I was able to figure out the issue with my set-up, so just an FYI for those in a similar situation.

Our set-up was using the `cimg/ruby:3.2.2-browsers` image. Apparently this image came packed with Chrome 116.0.5845.140. From looking at the output of the steps browser-tools sees Chrome 116.0.5845.140 exists and thus ignores the chrome-version 116.0.5845.96 set in the configuration.

When the ChromeDriver install step comes around, it'll then try to install 116.0.5845.140, which fails.

**Comment by BrianMitchL** [comment]
For `browser-tools/install-chrome`, I added `replace-existing: true` to force it to re-install and got it working for me. 

```yaml
      - browser-tools/install-chrome:
          chrome-version: 116.0.5845.96
          replace-existing: true
```

Looking at the docs, I think you'd need to use `replace-existing-chrome: true` for `browser-tools/install-browser-tools`

```yaml
      - browser-tools/install-browser-tools:
          chrome-version: 116.0.5845.96
          replace-existing-chrome: true
```

**Comment by perkinss** [comment]
The

> > > > > I was also able to workaround it with
> > > > > ```
> > > > >     steps:
> > > > >     - browser-tools/install-browser-tools:
> > > > >         chrome-version: 116.0.5845.96 # TODO: remove when chromedriver downloads are fixed
> > > > > ```
> > > > > 
> > > > > 
> > > > >     
> > > > >       
> > > > >     
> > > > > 
> > > > >       
> > > > >     
> > > > > 
> > > > >     
> > > > >   
> > > > > that also installs firefox and geckodriver, but doesn't need the apt get update at least
> > > > 
> > > > 
> > > > @hoylemd does this work for 1.4.4?
> > > 
> > > 
> > > It does in my project
> > 
> > 
> > Great. It also works for me. Thanks @hoylemd
> 
> This works for everything except MacOs. Looks like it ignores `chrome-version`

Same here, not using MacOS... but it's ignoring `chrome-version`. 

using 
```    steps:
    - browser-tools/install-browser-tools:
        chrome-version: 116.0.5845.96 # TODO: remove when chromedriver downloads are fixed
```
I get: 

```
Error getting version of chromedriver 116. Retrying with chromedriver 115 (attempt 1/5)
Error getting version of chromedriver 115. Retrying with chromedriver 114 (attempt 2/5)
Running tests in the browser...
SessionNotCreatedError: session not created: This version of ChromeDriver only supports Chrome version 114
Current browser version is 116.0.5845.140 with binary path /usr/bin/google-chrome
```

**Comment by perkinss** [comment]
> > > > > > I was also able to workaround it with
> > > > > > ```
> > > > > >     steps:
> > > > > >     - browser-tools/install-browser-tools:
> > > > > >         chrome-version: 116.0.5845.96 # TODO: remove when chromedriver downloads are fixed
> > > > > > ```
> > > > > > 
> > > > > > 
> > > > > >     
> > > > > >       
> > > > > >     
> > > > > > 
> > > > > >       
> > > > > >     
> > > > > > 
> > > > > >     
> > > > > >   
> > > > > > that also installs firefox and geckodriver, but doesn't need the apt get update at least
> > > > > 
> > > > > 
> > > > > @hoylemd does this work for 1.4.4?
> > > > 
> > > > 
> > > > It does in my project
> > > 
> > > 
> > > Great. It also works for me. Thanks @hoylemd
> > 
> > 
> > This works for everything except MacOs. Looks like it ignores `chrome-version`
> 
> Same here, not using MacOS... but it's ignoring `chrome-version`.
> 
> Going to try the `replace-existing: true` setting next

Nope, I am getting the following:
```
Error getting version of chromedriver 116. Retrying with chromedriver 115 (attempt 1/5)
Error getting version of chromedriver 115. Retrying with chromedriver 114 (attempt 2/5)
Running tests in the browser...
SessionNotCreatedError: session not created: This version of ChromeDriver only supports Chrome version 114
Current browser version is 116.0.5845.96 with binary path /usr/bin/google-chrome
...
```

**Comment by vietqhoang** [comment]
@perkinss 

It looks like you are using the `browser-tools/install-browser-tools` step.

If your step is using `browser-tools/install-browser-tools`, you'll need to use `replace-existing-chrome` instead of `replace-existing` for the configuration. Based on your comments, it seems you have tried using `replace-existing`.

The `replace-existing` is for the more defined install steps, like `browser-tools/install-chrome`

**Comment by perkinss** [comment]
> @perkinss
> 
> It looks like you are using the `browser-tools/install-browser-tools` step.
> 
> If your step is using `browser-tools/install-browser-tools`, you'll need to use `replace-existing-chrome` instead of `replace-existing` for the configuration. Based on your comments, it seems you have tried using `replace-existing`.
> 
> The `replace-existing` is for the more defined install steps, like `browser-tools/install-chrome`

I had updated the configuration to use install-chrome before using replace-existing:
```
    steps:
      - checkout
      - browser-tools/install-chrome:
          chrome-version: 116.0.5845.96 # TODO remove this when google updates the chrome driver
          replace-existing: true
```

**Comment by vietqhoang** [comment]
> Nope, I am getting the following:
> 
> ```
> Error getting version of chromedriver 116. Retrying with chromedriver 115 (attempt 1/5)
> Error getting version of chromedriver 115. Retrying with chromedriver 114 (attempt 2/5)
> Running tests in the browser...
> SessionNotCreatedError: session not created: This version of ChromeDriver only supports Chrome version 114
> Current browser version is 116.0.5845.96 with binary path /usr/bin/google-chrome
> ...
> ```

It looks like an older version of ChromeDriver is present from reading the output. Looks like your environment now has the desired version of Chrome.

Is there an install step for ChromeDriver after the Chrome install? From my understanding after reading [the documentation](https://circleci.com/developer/orbs/orb/circleci/browser-tools#commands-install-chromedriver) the ChromeDriver install will use the version of the installed Chrome. This tells me ChromeDriver install step needs to happen after Chrome.

I don't know how it'll behave with an existing ChromeDriver install.

**Comment by perkinss** [comment]
> It looks like an older version of ChromeDriver is present from reading the output. Looks like your environment now has the desired version of Chrome.
> 
> Is there an install step for ChromeDriver after the Chrome install? From my understanding after reading [the documentation](https://circleci.com/developer/orbs/orb/circleci/browser-tools#commands-install-chromedriver) the ChromeDriver install will use the version of the installed Chrome. This tells me ChromeDriver install step needs to happen after Chrome.
> 
> I don't know how it'll behave with an existing ChromeDriver install.

Oh! I thought it was attempting to install the chrome driver and not able to.

**Comment by jeremybdk** [comment]
@HrsUed maybe that could be a solution, can you share your config?

**Comment by UsAndRufus** [comment]
> For `browser-tools/install-chrome`, I added `replace-existing: true` to force it to re-install and got it working for me.
> 
> ```yaml
>       - browser-tools/install-chrome:
>           chrome-version: 116.0.5845.96
>           replace-existing: true
> ```
> 
> Looking at the docs, I think you'd need to use `replace-existing-chrome: true` for `browser-tools/install-browser-tools`
> 
> ```yaml
>       - browser-tools/install-browser-tools:
>           chrome-version: 116.0.5845.96
>           replace-existing-chrome: true
> ```

This worked for me, with an additional caveat. I had to pin the Chrome version to the version that the last passing test run used. We already had a step that printed out all tool versions, so it was easy to find. If I used the recommended version used here, our test suite failed.

**Comment by trkoch** [comment]
I'd like to reiterate this comment, which doesn't seem to get the attention it deserves.

As everybody closely following this issue probably noticed, we have a pretty clear case of [responsibility diffusion](https://en.wikipedia.org/wiki/Diffusion_of_responsibility), with A (CircleCI) pointing to B (Google) and B to A. In developer terms, an endless loop, or dependency conflict. There isn't going to be a solution any time soon. Considering that, please think twice if pinning a version of Chrome is a responsible thing to do. You know how this ends in two years' time when the upstream Chrome has diverged enough to make your tests fail in CI but not in development.

The `tldr;` is this:
* Update to Selenium `>=4.11`. This may also require updating parent dependencies such as `capybara`.
* Remove `browser-tools/install-chromedriver` from your `.circleci/config.yml`

Bonus: You benefit from automatic installation of `chromedriver` in your dev environment.

> If you use [selenium](https://github.com/SeleniumHQ/selenium) v4.11, perhaps you don't need to this Orb. Selenium v4.11 introduces a driver and browser manager which downloads browsers and drivers automatically. See https://www.selenium.dev/documentation/selenium_manager/#automated-browser-management

**Comment by ryanbourdais** [comment]
We had hoped that the versions would be synced by now, but as @hoylemd mentioned it has now been a week since this issue arose, and shows that this issue is not something that should just be worked around as it has had far longer effects than anticipated. As a result, I opened PR #94 to resolve this issue by checking for a 404 result from the generated chromedriver link and falling back to the highest chromedriver version that matches the major version of chrome installed.

**Comment by ryanbourdais** [comment]
I am going to cut a new patch orb version once that PR is merged in.

**Comment by ryanbourdais** [comment]
Browser Tools Orb version 1.4.5 is now cut and should resolve this issue. Please let me know if this issue persists or if any new issues pop up.

**Comment by 0xadada** [comment]
thanks @ryanbourdais, fixed for my team as well.

**Comment by karlentwistle** [comment]
I opened a new issue here https://github.com/CircleCI-Public/browser-tools-orb/issues/95

**Comment by ourmaninamsterdam** [comment]
`circleci/browser-tools@1.4.6` has resolved it for me

**Comment by berniechiu** [comment]
It happened again 

```
Chrome version major is 116
Installed version of Google Chrome is 116.0.5845.140 
500
116.0.5845.140 will be installed
/bin/bash: line 165: [[: 116.0.5845.140: syntax error: invalid arithmetic operator (error token is ".0.5845.140")
curl: (22) The requested URL returned error: 404
```

**Comment by iainbeeston** [comment]
@berniechiu There's a new GitHub issue for that https://github.com/CircleCI-Public/browser-tools-orb/issues/108

**Comment by jasonfb** [comment]
as of July 2023 I started getting this on my builds
```
Google Chrome 116.0.5845.140 is currently installed; replacing it
^@^@Preparing Chrome installation for Debian-based systems
https://dl.google.com/linux/chrome/deb/pool/main/g/google-chrome-stable/google-chrome-stable_116.0.5845.96-1_amd64.deb:
2024-07-19 15:41:42 ERROR 404: Not Found.
/bin/bash: line 130: google-chrome-stable: command not found
Google Chrome v116.0.5845.96 (stable) failed to install.

Exited with code exit status 1

```

**Comment by jasonfb** [comment]
the build failure started in July 2024 when I had the old browsertools (1.4.4) locked


this seems to be fixed by upgrading the browsertools and removing the special install stanzas

```
orbs:
  browser-tools: circleci/browser-tools@1.4.8
```

and then removing the re-installing stanzas:

(remove)
```
      - browser-tools/install-browser-tools:
          chrome-version: 116.0.5845.96 # TODO: remove when chromedriver downloads are fixed
          replace-existing-chrome: true
      - browser-tools/install-chrome:
          # TODO remove following line when fixed https://github.com/CircleCI-Public/browser-tools-orb/issues/90
          chrome-version: 116.0.5845.96
          replace-existing: true
```
